File: extend/task_relay/hub/context_crypto.py
"""Encrypt inline TaskSpec context at rest (M3 hardening)."""

from __future__ import annotations

import base64
import hashlib
import json
import os
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from extend.task_relay.hub.json_util import safe_json_loads

_ENVELOPE_KEY = "encrypted_inline"
_VERSION = 1


class ContextCryptoError(Exception):
    """Inline context encryption or decryption failed."""


def _derive_key(secret: str) -> bytes:
    if not secret:
        raise ContextCryptoError("encryption secret is required")
    return hashlib.sha256(secret.encode("utf-8")).digest()


def _encrypt_bytes(plaintext: bytes, secret: str) -> str:
    nonce = os.urandom(12)
    ciphertext = AESGCM(_derive_key(secret)).encrypt(nonce, plaintext, None)
    return base64.b64encode(nonce + ciphertext).decode("ascii")


def _decrypt_bytes(blob: str, secret: str) -> bytes:
    raw = base64.b64decode(blob.encode("ascii"))
    if len(raw) < 13:
        raise ContextCryptoError("ciphertext is too short")
    nonce, ciphertext = raw[:12], raw[12:]
    return AESGCM(_derive_key(secret)).decrypt(nonce, ciphertext, None)


def should_encrypt_context(context_json: str | None) -> bool:
    """Return True when context_json carries in-band inline payload."""
    if not context_json:
        return False
    payload = safe_json_loads(context_json) or {}
    return "inline" in payload or "inline_gzip" in payload


def encrypt_context_json(context_json: str, secret: str) -> str:
    """Wrap inline context in an encrypted envelope for storage."""
    if not should_encrypt_context(context_json):
        return context_json
    blob = _encrypt_bytes(context_json.encode("utf-8"), secret)
    return json.dumps({_ENVELOPE_KEY: {"v": _VERSION, "data": blob}})


def decrypt_context_json(context_json: str | None, secret: str) -> Any:
    """Return worker-ready context dict, decrypting stored envelopes when needed."""
    if not context_json:
        return None
    payload = safe_json_loads(context_json)
    if not isinstance(payload, dict):
        return payload
    envelope = payload.get(_ENVELOPE_KEY)
    if not isinstance(envelope, dict):
        return payload
    blob = envelope.get("data")
    if not isinstance(blob, str):
        raise ContextCryptoError("encrypted_inline.data is required")
    plaintext = _decrypt_bytes(blob, secret).decode("utf-8")
    return safe_json_loads(plaintext)

File: extend/task_relay/hub/go/internal/contextcrypto/contextcrypto.go
package contextcrypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
)

const envelopeKey = "encrypted_inline"

// Error indicates inline context encryption or decryption failed.
type Error struct {
	Msg string
}

func (e *Error) Error() string { return e.Msg }

func deriveKey(secret string) ([]byte, error) {
	if secret == "" {
		return nil, &Error{Msg: "encryption secret is required"}
	}
	sum := sha256.Sum256([]byte(secret))
	return sum[:], nil
}

func encryptBytes(plaintext []byte, secret string) (string, error) {
	key, err := deriveKey(secret)
	if err != nil {
		return "", err
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}
	ciphertext := gcm.Seal(nonce, nonce, plaintext, nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

func decryptBytes(blob, secret string) ([]byte, error) {
	key, err := deriveKey(secret)
	if err != nil {
		return nil, err
	}
	raw, err := base64.StdEncoding.DecodeString(blob)
	if err != nil {
		return nil, &Error{Msg: "invalid ciphertext encoding"}
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	if len(raw) < gcm.NonceSize() {
		return nil, &Error{Msg: "ciphertext is too short"}
	}
	nonce, ciphertext := raw[:gcm.NonceSize()], raw[gcm.NonceSize():]
	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, &Error{Msg: "decryption failed"}
	}
	return plaintext, nil
}

// ShouldEncrypt returns true when context JSON carries in-band inline payload.
func ShouldEncrypt(contextJSON string) bool {
	if contextJSON == "" {
		return false
	}
	var payload map[string]any
	if err := json.Unmarshal([]byte(contextJSON), &payload); err != nil {
		return false
	}
	_, hasInline := payload["inline"]
	_, hasGzip := payload["inline_gzip"]
	return hasInline || hasGzip
}

// EncryptContextJSON wraps inline context in an encrypted envelope for storage.
func EncryptContextJSON(contextJSON, secret string) (string, error) {
	if !ShouldEncrypt(contextJSON) {
		return contextJSON, nil
	}
	blob, err := encryptBytes([]byte(contextJSON), secret)
	if err != nil {
		return "", err
	}
	out, err := json.Marshal(map[string]any{
		envelopeKey: map[string]any{"v": 1, "data": blob},
	})
	if err != nil {
		return "", err
	}
	return string(out), nil
}

// DecryptContextJSON returns worker-ready context, decrypting stored envelopes.
func DecryptContextJSON(contextJSON, secret string) (any, error) {
	if contextJSON == "" {
		return nil, nil
	}
	var payload map[string]any
	if err := json.Unmarshal([]byte(contextJSON), &payload); err != nil {
		return nil, err
	}
	envelope, ok := payload[envelopeKey].(map[string]any)
	if !ok {
		return payload, nil
	}
	blob, _ := envelope["data"].(string)
	if blob == "" {
		return nil, &Error{Msg: "encrypted_inline.data is required"}
	}
	plaintext, err := decryptBytes(blob, secret)
	if err != nil {
		return nil, err
	}
	var decoded any
	if err := json.Unmarshal(plaintext, &decoded); err != nil {
		return nil, fmt.Errorf("decode decrypted context: %w", err)
	}
	return decoded, nil
}
